### PR TITLE
gh-100758: Refactor initialisation of frame headers into a single function (_PyFrame_Initialize)

### LINKED ISSUE #100758: Refactor initialisation of frame headers into a single function
There are 3 places in the code where we prepare a frame for execution:

1. _PyFrame_PushUnchecked in pycore_frame.h
2. init_frame in frameobject.c
3. _PyEvalFramePushAndInit in ceval.c

All of these call the function _PyFrame_InitializeSpecials to set values from the code object to various fields on the frame. Two of them (2 and 3 above) also NULL the localsplus array.

When we will change the frame layout for the register machine, currently we need to update those three places. But if we add NULLing the locals to _PyFrame_InitializeSpecials (and rename it to something clearer) then we can just make our changes there. 

There isn't really a problem with NULLing the localsplus in _PyFrame_PushUnchecked - all the call sites follow up by setting some fields and NULLing the rest. So now we will NULL all fields and overwrite a few after returning from _PyFrame_InitializeSpecials. Will benchmark to make sure.

(Lesson learned from the register machine experiment).

### Linked PRs
* gh-100759
* gh-100775

### PR DESCRIPTION
**Pull request by iritkatriel**
Fixes #100758.

This is version 1 (with an arg telling where to start NULLing).

* Issue: gh-100758

### CODE REVIEW

**Comment by carljm** [comment]
This looks OK to me, pending benchmarking. It means we will do one or two extra NULL writes to localsplus entries that are about to get written again with a value, in some hot paths that use . (This could be avoided with an extra integer argument.) Will wait and see what Mark thinks.

File: Include/internal/pycore_frame.h [hunk 1]
```diff
@@ -101,7 +101,7 @@ void _PyFrame_Copy(_PyInterpreterFrame *src, _PyInterpreterFrame *dest);
    when frame is linked into the frame stack.
  */
 static inline void
-_PyFrame_InitializeSpecials(
+_PyFrame_InitializeHeader(
```
Comment (markshannon) [inline Include/internal/pycore_frame.h]: If you initialize the locals, then this initializes the whole frame. Maybe call it `_PyFrame_Initialize`?

File: Include/internal/pycore_frame.h [hunk 2]
```diff
@@ -115,6 +115,11 @@ _PyFrame_InitializeSpecials(
     frame->prev_instr = _PyCode_CODE(code) - 1;
     frame->yield_offset = 0;
     frame->owner = FRAME_OWNED_BY_THREAD;
+
+    PyObject **localsarray = &frame->localsplus[0];
```
Comment (markshannon) [inline Include/internal/pycore_frame.h]: Any compiler will do loop invariant code motion, so no need to move this out of the loop.

**Review by markshannon — COMMENTED** [review]
I like @carljm 's suggestion of passing in the index of the first local to NULL out.

I'd like to see the benchmarks.
If benchmarking shows no difference, then go with this approach as it is the cleanest.

File: Python/bytecodes.c [hunk 3]
```diff
@@ -2818,7 +2806,7 @@ dummy_func(
             DEOPT_IF(argcount < minargs, CALL);
             DEOPT_IF(!_PyThreadState_HasStackSpace(tstate, code->co_framesize), CALL);
             STAT_INC(CALL, hit);
-            _PyInterpreterFrame *new_frame = _PyFrame_PushUnchecked(tstate, func);
+            _PyInterpreterFrame *new_frame = _PyFrame_PushUnchecked(tstate, func, argcount);
```
Comment (iritkatriel) [inline Python/bytecodes.c]:
```suggestion
            _PyInterpreterFrame *new_frame = _PyFrame_PushUnchecked(tstate, func, code->co_argcount);
```

**Comment by iritkatriel** [comment]
These are benchmarking results from before I changed to pass the int arg: 
https://github.com/faster-cpython/benchmarking/blob/main/results/bm-20230104-3.12.0a3%2B-e73d0cf/bm-20230104-linux-x86_64-iritkatriel-InitializeHeader-3.12.0a3+-e73d0cf-vs-base.md

(overall x0 faster).

I will rerun with the last change.

**Review by carljm — APPROVED** [review]
Looks great to me, presuming the second benchmark run will also show no impact.

**Comment by iritkatriel** [comment]
The results for this version are here: https://github.com/faster-cpython/benchmarking/blob/main/results/bm-20230105-3.12.0a3%2B-d501577/bm-20230105-linux-x86_64-iritkatriel-InitializeHeader-3.12.0a3+-d501577-vs-base.md

0x00 slower.

**Comment by markshannon** [comment]
Might as well go with the faster version.

**Comment by iritkatriel** [comment]
You don't think this is noise?

**Comment by iritkatriel** [comment]
I created https://github.com/python/cpython/pull/100775 with the arg removed, will run the benchmark one more time.

**Comment by iritkatriel** [comment]
The result of the last run are here: https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20230105-3.12.0a3%2B-94fa28c/bm-20230105-linux-x86_64-iritkatriel-InitializeHeader_noa-3.12.0a3+-94fa28c-vs-base.md

**Comment by markshannon** [comment]
I'm merging this, as the performance difference with this and #100775 is in the noise, and this seems the more efficient design.

**Pull request merged**